### PR TITLE
Add an autopkgtest boot test

### PR DESCRIPTION
**Pull request by warsaw**
[LP: #1625732](https://bugs.launchpad.net/ubuntu-image/+bug/1625732)

This test creates an amd64 image with an extra snap that implements a "case swapping echo service" that runs at boot time.  Then it boots the image and tries to talk to the echo service port.  If you get back what you sent it (case swapped), then we know everything worked.

### CODE REVIEW

**Comment by warsaw** [comment]
@vorlonofportland @sil2100 Pending CI passing, this is ready if you guys want to review it.

File: debian/tests/boot [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        # We can't but it makes no sense to fail.
+        sys.exit(0)
+    image = build_image()
+    with ExitStack() as resources:
+        thread = Thread(target=qemu, args=(image,))
+        thread.daemon = True
+        thread.start()
+        # Connect to the echo service.  There are lots of factors for why this
+        # can be slow and/or temporarily fail.  Basically, the echo-service
+        # snap won't start responding until the bring-up-network snap succeeds
```
Comment (mwhudson) [inline debian/tests/boot]: I didn't think there was a bring-up-snap any more?

Comment (warsaw) [inline debian/tests/boot]: Thanks!  I'll fix that comment.

File: debian/tests/boot [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+from socket import create_connection
+from subprocess import run
+from tempfile import TemporaryDirectory
+from threading import Thread
+
+
+TMP = os.environ['AUTOPKGTEST_TMP']
+DIR = os.path.abspath(os.path.join('debian', 'tests', 'models'))
+ECHO_TIMEOUT = 60 * 5 # the VM can be slow
+NEWLINE = b'\n'
```
Comment (sil2100) [inline debian/tests/boot]: I don't see this global NEWLINE constant used anywhere, is it still needed?

File: debian/tests/boot [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    return images[0]
+
+
+def main():
+    # First off, can we even run the test on this architecture?
+    if platform.machine() != 'x86_64':
+        # We can't but it makes no sense to fail.
+        sys.exit(0)
+    image = build_image()
+    with ExitStack() as resources:
```
Comment (sil2100) [inline debian/tests/boot]: I'm still a newb regarding using ExitStack(), but is it necessary here? I don't see 'resources' being used anywhere (no cleanup or anything) so just wonder if it's required.

File: debian/tests/boot (hunk 2 above)
Comment (warsaw) [inline debian/tests/boot]: Good catch, it's not!  Thanks.

File: debian/tests/boot (hunk 3 above)
Comment (warsaw) [inline debian/tests/boot]: Indeed, it's not.  I just let the autopkgtest infra tear everything down, so no need to explicitly collect resources.  Thanks, good catch.

**Pull request merged**